[BEAM-3942] Update performance testing framework to use Gradle

lgajowy opened the pull request:
This PR introduces three gradle tasks that are needed for IOITs. 

1. integrationTest - which is for running integration tests manually
2. packageIntegrationTests - which provides necessary IOIT's dependencies (runners to be chosen using a command line arg)
3. performanceTest - which runs perfkit to setup all necessary testing infrastructure. 

At the time of writing this mesage, PerfKitBenchmarker is not using gradle yet. I'll migrate it as soon as this PR gets merged. More on that in the issues comments: https://issues.apache.org/jira/browse/BEAM-3942 

@jbonofre could you take a look?

CC: @lukecwik 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

lgajowy commented:
Run seed job

lgajowy commented:
Sorry, the seed job i run does not make sens for now - jenkins (perfkit) still uses maven (yet).

lgajowy commented:
Also cc: @chamikaramj

lukecwik commented:
R: @udim

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+class JavaPerformanceTestConfiguration {
+  String pkbLocation = System.getProperty('pkbLocation')
+
+  String logLevel = System.getProperty('logLevel', 'INFO')
```
Could you add single lines of comments explaining what these properties are ? For example, I'm not sure what 'official' is used for.

lgajowy replied:
ok

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+  String beamPrebuilt = System.getProperty('beamPrebuilt', 'true')
+  String beamSdk = System.getProperty('beamSdk', 'java')
+
+  String timeout = System.getProperty('itTimeout', '1200')
```
Do all ITs have to have to same timeout

lgajowy replied:
No, the timeout here is default and can be changed in case someone wants to set a different one. In most cases 1200 sec will be enough.

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+  String kubeconfig = System.getProperty('kubeconfig', System.getProperty('user.home') + '/.kube/config')
+  String kubectl = System.getProperty('kubectl', 'kubectl')
+
+  // PerfKit will have trouble reading 'null' path. It expects empty string if no scripts are expected.
```
Is this something we can fix at PerfKit side (and not specify empty strings here) ?

lgajowy replied:
I think we can fix this on the perfkit side but I rather tend to think it should be done in separate PR. I wanted to change the least possible amount of code in perfkit to integrate with gradle and then improve. Do you agree with me?

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+
+  String integrationTestPipelineOptions = System.getProperty('integrationTestPipelineOptions')
+
+  // PerfKit will have trouble reading 'null' path. It expects empty string if no config file is expected.
```
Ditto.

lgajowy replied:
(answered above)

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+  String extraProperties = System.getProperty('beamExtraProperties', '')
+}
+
+// Configures a project with a set of tasks needed for running performance tests
```
What is a 'project' here ?

lgajowy replied:
I will fix this doc to be more clear

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+ext.enableJavaPerformanceTesting = {
+  println "enableJavaPerformanceTesting with ${it ? "$it" : "default configuration"} for project ${project.name}"
+
+  // Use the implicit it parameter of the closure to handle zero argument or one argument map calls.
```
implicit 'it' parameter.

BTW who sets this parameter ?

lgajowy replied:
Provided a link that clarifies how to use the implicit 'it' parameter.

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+
+  // Add runners needed to run integration tests on
+  task packageIntegrationTests(type: Jar) {
+    if (gradle.startParameter.taskNames.contains('integrationTest')) {
```
So, tasks names starting with 'integrationTest' is a requirement ?

lgajowy replied:
This line effectively means that we should include the runner dependencies only when the `integrationTest` task is run. So when user runs `./gradlew integrationTest...` those dependencies are added while building the jar. 

Now I see that this is not enough due the following reason: there is no way to "prebuild" tests - we cannot build with a separate build command and then just run the tests. Such prebuilding is needed by perfkit so it's a necessity I forgot to take into account here. "Prebuilding" is invoked when we specify the "beam_prebuilt" flag to false, see more in [Perfkit's code](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/a3e6c0382809cb68cd294f34cbd5b3b558e53f56/perfkitbenchmarker/beam_benchmark_helper.py#L147).

I think we should remove this condition and include the runner dependency only if the command is run with appropriate "integrationTestRunner" property . I changed this in the next commit. 
So now we can prebuild in Perfkit using this command: 
`./gradlew clean build -xtest` -DintegrationTestRunner=dataflow 
and run the tests using `integrationTest` task, for example like this: 

and execute the tests using this one:
`./gradlew clean integrationTest -p sdks/java/io/file-based-io-tests/ -DintegrationTestPipelineOptions='[]' --tests org.apache.beam.sdk.io.text.TextIOIT  -DintegrationTestRunner=dataflow`

WDYT?

aaltay commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+
+  String logLevel = System.getProperty('logLevel', 'INFO')
+  String gradleBinary = System.getProperty('gradleBinary', './gradlew')
+  String isOfficial = System.getProperty('official', 'true')
```
Do we want to default to `true` here?

lgajowy replied:
No we shouldn't - this would mark benchmark results being marked as official which even in Perfkit is defaulted to false. The task is for running them from local machines so those are definitely not official test results.  It should be false then - my bad.

I added a comment describing this in more detail in next commit

aaltay commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+
+      def runner = configuration.runner
+      dependencies {
+        if (runner.contains('dataflow')) {
```
Do we need to add hard coded branch for each possible runner here?

lgajowy replied:
Currently we run the performance test only on dataflow runner. No other runner is supported, eg. due to issues like this: 

https://issues.apache.org/jira/browse/BEAM-3359
https://issues.apache.org/jira/browse/BEAM-3370
https://issues.apache.org/jira/browse/BEAM-3371

We didn't try to run ITs on runners like apex or gearpump so we don't know whether ITs run there.

This is why I decided not to add them. I think we should add new runner dependencies here only when they are ready to run ITs. Wdyt?

aaltay replied:
This is fine. I was asking about, if there is a more scalable way of adding new runner dependencies here in the future. (As opposed to bunch if this runner ... lines.)

aaltay commented on sdks/java/io/cassandra/build.gradle:
```diff
@@ -18,6 +18,7 @@
 
 apply from: project(":").file("build_rules.gradle")
 applyJavaNature(artifactId: "beam-sdks-java-io-cassandra")
+enableJavaPerformanceTesting()
```
What does this line do? Does it enable for all build?

lgajowy replied:
This adds `integrationTest` task to cassandra io's build.gradle file. It works the same way as `applyJavaNature()` but adds integration testing related things. The way it is done now requires developers explicitly "append" the tasks to every module that needs ITs by adding `enableJavaPerformanceTesting()` in build.gradle file.

An alternative would be to include this task (along with `packageIntegrationTests` ) to `applyJavaNature()` closure. This would add the tasks (`integrationTests`, `packageIntegrationTest` to every module that uses `applyJavaNature()`, so it's more implicit. 

What do you think? Which option do you like better?

aaltay replied:
The current approach seems better. Thank you for the explanation.

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+
+      def runner = configuration.runner
+      dependencies {
+        if (runner.contains('dataflow')) {
```
Why do we have to set runner-specific dependencies here? Can't this be set by the user command that invokes the Gradle task to run the test ?

lgajowy replied:
It mimcs the maven behaviour that was there + is implemented in one generic place (code is not duplicated all over the build configurations as it was in maven). Because we don't have profiles in gradle, a system property was used to decide which runner to include - this is done by user by specifying `-DintegrationTestRunner=<runner_to_include>`

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+  println "createPerformanceTestHarness with ${it ? "$it" : "default configuration"} for project ${project.name}"
+
+  // Use the implicit it parameter of the closure to handle zero argument or one argument map calls.
+  JavaPerformanceTestConfiguration configuration = it ? it as JavaPerformanceTestConfiguration : new JavaPerformanceTestConfiguration()
```
Ditto. Please explain what 'it' parameter is.

lgajowy replied:
ok

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+  // There are more options with default values that can be tweaked if needed (see below).
+  task performanceTest(type: Exec) {
+
+    // Perfkit needs to work in the Beam's root directory,
```
You mean, test has to be invoked from the root directory ?

lgajowy replied:
I made this comment more clear in the next commit too. I meant that the performanceTest task requires perfkit to be run in the project root directory.

chamikaramj commented on sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java:
```diff
@@ -47,18 +47,16 @@
  * </pre>
  *
- * <p>If you want to run this with a runner besides directrunner, there are profiles for dataflow
- * and spark in the jdbc pom. You'll want to activate those in addition to the normal test runner
```
Does this restriction not apply anymore ?

lgajowy commented:
@aaltay @chamikaramj thank you for all your comments and suggestions! fyi, please expect my answers on tuesday (I'm a little bit unavailable today).

lgajowy commented on build.gradle:
```diff
@@ -22,6 +22,7 @@
 // tasks.
 
 apply from: project(":").file("build_rules.gradle")
+createPerformanceTestHarness()
```
ok

lgajowy commented:
Thank you guys for all the comments. I responded to them and created another commit with changes, mostly in documentation to ease your effort and make things more clear. 

@chamikaramj There already is a documentation page describing how [the framework](https://beam.apache.org/documentation/io/testing/#i-o-transform-integration-tests) works. Part of the gradle migration task is to update it to Gradle. So clear description will be available soon.Meanwhile I created a [Performance testing framework documentation errata](https://docs.google.com/document/d/1CJZURnqCabc_GA-qgC3JYKQdhh00T2l5FzFS0DF12wU/edit?usp=sharing) that provides updated commands and (possibly in the future) more things to be changed in the docs This will also make getting the documentation up to date easier. Of course feel free to add comments and suggestions. Thanks!

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,149 @@ ext.applyJavaNature = {
+}
+
+// When applied in a module's build.gradle file, this closure provides set of tasks
+// needed to run IOIT integration tests (manually, without perfkit).
```
<!--new_thread; commit:a2d1efe7c481a850f0e744ee6d4372c885c695f8; resolved:0-->
s/perfkit/PerfKitBenchmarker (here and elsewhere).

lgajowy replied:
ok

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,149 @@ ext.applyJavaNature = {
+  String optionsConfigFile = System.getProperty('beamITOptions', '')
+
+  // Fully qualified name of the test to be run, eg:
+  // 'org.apache.beam.sdks.java.io.jdbc.JdbcIOIT'.
```
<!--new_thread; commit:a2d1efe7c481a850f0e744ee6d4372c885c695f8; resolved:0-->
Can we group required and optional properties separately ?

lgajowy replied:
ok

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,149 @@ ext.applyJavaNature = {
+  // Specifies names of benchmarks to be run by perfkit.
+  String benchmarks = System.getProperty('benchmarks', 'beam_integration_benchmark')
+
+  // If beam is not "prebuilt then perfkit runs the build task before running the tests.
```
<!--new_thread; commit:a2d1efe7c481a850f0e744ee6d4372c885c695f8; resolved:0-->
Nit: end quote after prebuilt.

lgajowy replied:
ok

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,149 @@ ext.applyJavaNature = {
+  }
+
+  // Task for running integration tests
+  task integrationTest(type: Test) {
```
<!--new_thread; commit:a2d1efe7c481a850f0e744ee6d4372c885c695f8; resolved:0-->
Why do we have to define a new task for integration tests ? I think for Maven we used a standard task (verify). Is this a requirement for Gradle ?

lgajowy replied:
I wouldn't call this a requirement - rather a good gradle practice to handle things this way. 

For invoking the integration test maven used the Failsafe plugin. It was designed in a way to look for all test classes that name ends with "IT" and execute them. On the contrary unit tests were run using Surefire plugin. They both "connected" to mavens "verify" phase and this was a problem in maven: normally when someone typed `mvn verify` all plugins were fired. This is why we needed to "skipITs" or skip unit tests depending on which kind of tests we were going to execute and introduce profiles like "io-it" for this job: https://github.com/apache/beam/blob/b8cf07dc03643ded1d4e9d45caf4151316cf2c3e/sdks/java/io/pom.xml#L104

Gradle is built totally differently there and has no phases. It operates on independent tasks each representing "an atomic piece of work". This is IMO a perfect fit for our case: we define a Test type task called `integrationTest` or `performanceTest` and invoke it on demand, not worrying that unit tests will fire along with integration tests. Thanks to that we do not have to define a whole series of additional profiles. It's sufficient to have just two tasks defined in one place.

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+  String kubeconfig = System.getProperty('kubeconfig', System.getProperty('user.home') + '/.kube/config')
+  String kubectl = System.getProperty('kubectl', 'kubectl')
+
+  // PerfKit will have trouble reading 'null' path. It expects empty string if no scripts are expected.
```
<!--thread_id:cc_179905431_t; commit:eebddd5f85b675523294b22a3efe8e110c8d145e; resolved:1-->
<!--section:context-quote-->
> **lgajowy** wrote:
> I think we can fix this on the perfkit side but I rather tend to think it should be done in separate PR. I wanted to change the least possible amount of code in perfkit to integrate with gradle and then improve. Do you agree with me?

<!--section:body-->
Sounds good.

chamikaramj commented on build_rules.gradle:
```diff
@@ -593,6 +593,118 @@ ext.applyJavaNature = {
+
+  // Add runners needed to run integration tests on
+  task packageIntegrationTests(type: Jar) {
+    if (gradle.startParameter.taskNames.contains('integrationTest')) {
```
<!--thread_id:cc_179905518_t; commit:eebddd5f85b675523294b22a3efe8e110c8d145e; resolved:1-->
<!--section:context-quote-->
> **lgajowy** wrote:
> This line effectively means that we should include the runner dependencies only when the `integrationTest` task is run. So when user runs `./gradlew integrationTest...` those dependencies are added while building the jar. 
> 
> Now I see that this is not enough due the following reason: there is no way to "prebuild" tests - we cannot build with a separate build command and then just run the tests. Such prebuilding is needed by perfkit so it's a necessity I forgot to take into account here. "Prebuilding" is invoked when we specify the "beam_prebuilt" flag to false, see more in [Perfkit's code](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/a3e6c0382809cb68cd294f34cbd5b3b558e53f56/perfkitbenchmarker/beam_benchmark_helper.py#L147).
> 
> I think we should remove this condition and include the runner dependency only if the command is run with appropriate "integrationTestRunner" property . I changed this in the next commit. 
> So now we can prebuild in Perfkit using this command: 
> `./gradlew clean build -xtest` -DintegrationTestRunner=dataflow 
> and run the tests using `integrationTest` task, for example like this: 
> 
> and execute the tests using this one:
> `./gradlew clean integrationTest -p sdks/java/io/file-based-io-tests/ -DintegrationTestPipelineOptions='[]' --tests org.apache.beam.sdk.io.text.TextIOIT  -DintegrationTestRunner=dataflow`
> 
> WDYT?

<!--section:body-->
Makes sense. Thanks.

lgajowy commented on build_rules.gradle:
```diff
@@ -837,6 +837,154 @@ ext.applyJavaNature = {
+  String extraProperties = System.getProperty('beamExtraProperties', '')
+
+  // Runner which will be used for running the tests.
+  String runner = System.getProperty('integrationTestRunner', 'direct')
```
I think, unfortunately I've made a mistake here: we shouldn't set default value of this property. This will result in adding direct runner as a dependency everytime building is run (even if we are not running the integrationTest task). Sorry for that!

A solution would be to not set the default value - the dependency won't be included when there is no runner to attach. WDYT?

lgajowy replied:
I think, unfortunately I've made a mistake here: we shouldn't set default value of this property. This will result in adding direct runner as a dependency everytime building is run (even if we are not running the integrationTest task). Sorry for that!

A solution would be to not set the default value - the dependency won't be included when there is no runner to attach. This property would then be "always required" WDYT?

lgajowy commented on build_rules.gradle:
```diff
@@ -837,6 +837,154 @@ ext.applyJavaNature = {
+  /* Always required properties: */
+
+  // Pipeline options to be used by the tested pipeline.
+  String integrationTestPipelineOptions = System.getProperty('integrationTestPipelineOptions')
```
I think, unfortunately I've made a mistake here: we shouldn't set default value of this property. This will result in adding direct runner as a dependency everytime building is run (even if we are not running the integrationTest task). Sorry for that!

A solution would be to not set the default value - the dependency won't be included when there is no runner to attach. This property would then be "always required" WDYT?

lgajowy commented:
Please don't merge this yet.

We already started running tests on hdfs filesystem on Jenkins while this PR was being rewiewed. Therefore, support for filesystems needs to be done to be able to run all current Jenkins performance tests. Sorry for not noticing this earlier. I'll prepare a fix shortly. It will be very similar to runner support: dependencies needed by certain filesystems (eg. hdfs) can be included conditionally based on a "filesystem" property.

lgajowy commented:
Update: I was able to successfully run tests on hdfs filesystem too using "integrationTest" task but there are several issues due to previous merge commit that prevent me from running the performanceTest task. Working on demystifying and solving the issue.

lgajowy commented:
Things that changed (see the commit descriptions for more details): 

1. Filesystem support was added (crucial for recently added hdfs IOITs)
2. The direct runner dependency was deleted from file-based-io-tests module (it is added either way by "packageIntegrationTests" task and caused problems (couldn't compile)
3. Moved _integrationTest_ and _packageIntegrationTests_ tasks to different closures to make integration test configuration more flexible and avoid cyclic dependencies in some cases. 

Due to some problems i had after merging master recently to this branch, I also made sure that all configurations needed by current Jenkins jobs work. I also updated the "[errata](https://docs.google.com/document/d/1CJZURnqCabc_GA-qgC3JYKQdhh00T2l5FzFS0DF12wU/edit)" document. 

Could you guys take a look again?

chamikaramj commented on build_rules.gradle:
```diff
@@ -837,6 +837,179 @@ ext.applyJavaNature = {
+  //  -DintegrationTestPipelineOptions='["--numberOfRecords=1000", "<more options>"]' \
+  //  -DintegrationTest=<io test, eg. org.apache.beam.sdk.io.text.TextIOIT> \
+  //  -DitModule=<directory containing desired test, eg. sdks/java/io/file-based-io-tests> \
+  //  -DintegrationTestRunner=direct/dataflow/<other>
```
<!--new_thread; commit:3165e6e7301ddfb1565b1de36397bfac474d253b; resolved:0-->
What's the value in providing multiple values for 'runner' parameter. Don't we always run a single execution using a single runner ?

lgajowy replied:
There is no value at all - it's only my inadvertence. :) Thanks to your comment I realised this may be misguiding. I'll rephrase this.

chamikaramj commented on build_rules.gradle:
```diff
@@ -837,6 +837,179 @@ ext.applyJavaNature = {
+    dependencies {
+
+      /* include dependencies required by runners */
+      if (runner?.contains('dataflow')) {
```
<!--new_thread; commit:3165e6e7301ddfb1565b1de36397bfac474d253b; resolved:0-->
Do we have to ignore case ?

lgajowy replied:
I think we should not be case sensitive here because it will be easier to use (eg. dataflow and DataFlow is still the same runner).  We should use `equalsIgnoreCase()` instead of contains, because there's no sense to provide multiple runners/filesystems. I'll fix this too, thanks.

chamikaramj commented on build_rules.gradle:
```diff
@@ -837,6 +837,179 @@ ext.applyJavaNature = {
+      }
+
+      /* include dependencies required by filesystems */
+      if (filesystem?.contains('hdfs')) {
```
<!--new_thread; commit:3165e6e7301ddfb1565b1de36397bfac474d253b; resolved:0-->
Ditto.

lgajowy replied:
(answered above)

lgajowy commented:
Done. Rebased to master and divided into two commits.

lgajowy commented:
retest this please

chamikaramj commented:
Retest this please

chamikaramj commented:
LGTM
